assets/src/edit-story/app/history/test/reducer.js

```diff
@@ -0,0 +1,273 @@
+    expect(() =>
+      reducer(EMPTY_STATE, {
+        type: 'UNKNOWN_ACTION',
+        payload: {},
+      })
+    ).toThrow(/Unknown history reducer action: UNKNOWN_ACTION/i);
+  });
+
+  describe('addEntry', () => {
+    it('should add entry to the beginning of the history', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }],
+      };
+
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries[0]).toStrictEqual(newEntry);
+    });
+
+    it('should clear undone states when adding a new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }, { id: 2 }, { id: 3 }],
+        offset: 2,
+      };
+      const newEntry = { id: 4 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries).toHaveLength(2);
+      expect(result.entries[1].id).toStrictEqual(3);
+      expect(result.entries[0].id).toStrictEqual(4);
+    });
+
+    it('should set correct state values when adding new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        versionNumber: 3,
+        entries: [{ id: 1 }],
+        offset: 2,
+      };
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.versionNumber).toStrictEqual(4);
+      expect(result.offset).toStrictEqual(0);
+    });
+
+    it('should respect the determined history size when adding a new entry', () => {
+      const entries = [...Array(size + 1)].map(() => {
+        return { id: 'foo' };
+      });
+      const initialState = {
+        ...EMPTY_STATE,
+        entries,
+      };
+      const newEntry = { id: 7 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+      expect(result.entries).toHaveLength(size);
+    });
+
+    it('should set the correct offset when replaying', () => {
+      const requestedState = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 2 }, requestedState],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.offset).toStrictEqual(1);
+    });
+
+    it('should clear the requested state after replaying', () => {
+      const requestedState = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 2 }, requestedState],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.requestedState).toBeNull();
+    });
```

Comment:
I'd combine this to a single test - and also in that test verify, that the list of entries is not growing (the replayed state is not added as a new state).